_drafts/postgres-operations-that-you-cant-explain.md

```diff
@@ -0,0 +1,282 @@
+---
+layout: post
+title: "Postgres operations that you can't EXPLAIN"
+---
+
+I am currently dabbling in postgres internals in my spare time, including
+trying to understand the query planner and executor. It's a deeply humbling
+experience, but occasionally I'm delighted to be able to answer simple
+questions I'm researching.
+
+One of these questions is how postgres is executing `ORDER BY` clauses inside
+of aggregate calls.
+
+Consider a simple table made of 3 integers, 1...3:
+
+```
+postgres=# CREATE TABLE foo AS
+postgres-# SELECT * FROM generate_series(1, 3) i;
+SELECT 3
+postgres=# SELECT * FROM foo;
+ i
+---
+ 1
+ 2
+ 3
+(3 rows)
+```
+
+Now let's assume we want to convert those rows into a postgres array in
+descending order:
+
+```
+postgres=# SELECT array_agg(i ORDER BY i DESC) FROM foo;
```

Comment:
Here's a thought... maybe you could start simpler here and then explain the syntax? I'm thinking something like this:

> Now let's assume we want to convert those rows into a postgres array. We'd use the `array_agg` aggregate function like this:
> `SELECT array_agg(i) FROM foo;`
> If we now want to sort the elements in the array, let's say in descending order, we can use the [order-by-clause support](https://www.postgresql.org/docs/9.5/static/sql-expressions.html#SYNTAX-AGGREGATES) of the `array_agg` function:
> `SELECT array_agg(i ORDER BY i DESC) FROM foo;`

Just a thought, because I didn't know about this (awesome!) feature.